ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -227,123 +316,154 @@ prune maxSize st
---
---   > rewind (lastSlot st) st == Just st
---
--- * Rewinding to the same slot should be idempotent:
---
---   > rewind s st == Just st'  ==>  rewind s st' == Just st'
+-- implicitly applied).
 --
--- * Rewinding fails only if the slot is before the anchor
---
---   > rewind s st == Nothing  iff  s < anchor st
---
---   (See documentation of 'anchor' for motivation)
---
--- * If rewinding to a particular slot fails, rewinding to any slot prior to
---   that will also fail
---
---   > rewind s st == Nothing, s' < s  ==>  rewind s' st == Nothing
---
--- * Rewinding to the anchor will leave the state empty
---
---   > size (rewind (anchor st) st) == 0
---
--- * But rewinding to a more recent block will not
---
---   > s > anchor st  ==>  size (rewind s st) > 0
-rewind :: PBftCrypto c
-       => WithOrigin SlotNo -> PBftChainState c -> Maybe (PBftChainState c)
-rewind slot st
-  | slot < anchor st = Nothing
-  | otherwise = Just $ PBftChainState {
-        anchor  = anchor st
-      , signers = keep
-      -- Typically we only drop a few blocks (and keep most), so updating
-      -- the counts based on what we drop will in most cases be faster than
-      -- recomputing it from the blocks we keep.
-      , counts  = repeatedly (decrementKey . pbftSignerGenesisKey)
-                             (toList discard)
-                             (counts st)
-      }
+-- In addition to preserving the invariant, we also have the guarantee that
+-- rolling back to a point (within @k@) and then reapplying the blocks that were
+-- rolled back results in the original state.
+rewind :: forall c. PBftCrypto c
+       => SecurityParam
+       -> WindowSize
+       -> WithOrigin SlotNo
+       -> PBftChainState c -> Maybe (PBftChainState c)
+rewind k n mSlot PBftChainState{..} =
+    case mSlot of
+      At slot ->
+        -- We scan from the right, since block to roll back to likely at end
+        case Seq.spanr (\(PBftSigner slot' _) -> slot' > slot) postAnchor of
+
+          -- We found the slot to roll back to post-anchor. Discard everything
+          -- after that slot.
+          (toDiscard, toKeep@(_ :|> x)) ->
+            if slot == pbftSignerSlotNo x
+              then Just $ go toDiscard toKeep
+              else notPreviouslyApplied
+
+          -- The slot was not found post-anchor. If the slot matches the last
+          -- slot pre-anchor, all is well, discarding everything post-anchor.
+          -- Otherwise, the rollback is too far.
+          (toDiscard, Empty) ->
+            case preAnchor of
+              _ :|> x
+                | slot == pbftSignerSlotNo x -> Just $ go toDiscard Empty
+                | slot <  pbftSignerSlotNo x -> rollbackToFar
+                | otherwise                  -> notPreviouslyApplied
+              _otherwise ->
+                notPreviouslyApplied
+
+      -- We can only roll back to origin if there are no signatures pre-anchor.
+      Origin ->
+        case preAnchor of
```

Comment:
Comment: Rolling back to the origin would leave the chain empty. This is only possible if we have at most `k` blocks in the chain. If we have more than `k` blocks, the pre-anchor will not be empty. 

TODO: Check if the tests cover this case.

Reply:
They do.